### PR TITLE
FIX OptionsetField returns valid HTML #4901

### PR DESCRIPTION
**Pull request by torleif**

### CODE REVIEW

File: forms/OptionsetField.php [hunk 1]
```diff
@@ -102,4 +102,19 @@ public function validate($validator) {
...
+	public function getAttributes() {
+		$attributes = parent::getAttributes();
+		unset($attributes['name']);
+		if(isset($attributes['required'])) {
+			unset($attributes['required']);
+		}
+		if(isset($attributes['role'])) {
+			unset($attributes['role']);
+		}
+		if(isset($attributes['aria-required'])) {
```
Comment (tractorcow) [inline forms/OptionsetField.php]: How about doing something like: http://stackoverflow.com/questions/34300154/is-aria-required-allowed-on-radio-buttons-or-checkboxes

We'd need to put the options into a fieldset, though.

File: forms/OptionsetField.php [hunk 2]
```diff
@@ -102,4 +102,19 @@ public function validate($validator) {
 	public function ExtraOptions() {
 		return new ArrayList();
 	}
+	
+	public function getAttributes() {
+		$attributes = parent::getAttributes();
+		unset($attributes['name']);
+		if(isset($attributes['required'])) {
+			unset($attributes['required']);
```
Comment (tractorcow) [inline forms/OptionsetField.php]: You don't need to check isset() before doing unset(). It's a safe operation. :)

File: forms/OptionsetField.php (hunk 1 above)
Comment (torleif) [inline forms/OptionsetField.php]: could we could do something smart, like checking if there is a fieldset as a parent and then set the aria-required on it?

Comment (tractorcow) [inline forms/OptionsetField.php]: I think if we build it in, it'll probably not need the check.


```
<fieldset $AttributesHTML>
    <ul>
	<% loop $Options %>
		<li class="$Class">
			<input id="$ID" class="radio" name="$Name" type="radio" value="$Value"<% if $isChecked %> checked<% end_if %><% if $isDisabled %> disabled<% end_if %> />
			<label for="$ID">$Title</label>
		</li>
	<% end_loop %>
</ul>
</fieldset>
```

Comment (tractorcow) [inline forms/OptionsetField.php]: Is it semver-safe to wrap a HTML element in a fieldset? Maybe a front-ender could help answer that question. :) @jonom ?

Comment (torleif) [inline forms/OptionsetField.php]: On second thought checking the parent would be a dumb idea, as the parent fieldset container could contain non required elements.

Comment (torleif) [inline forms/OptionsetField.php]: looking into this a little deeper, according to the w3 spec: *Attribute aria-required is unnecessary for elements that have attribute required.*

So aria-required seems to be now superfluous. thus the simple fix here: https://github.com/torleif/silverstripe-framework/commit/31fe519ec3b3d7d05cc422e7f574552127f7e007

should be able to resolve the required issue 

Note the required element only has to be on the first element, and I've done it for all. I could add a <% if First %> ... on there too.

Comment (tractorcow) [inline forms/OptionsetField.php]: I've done a bit of digging, and it seems (at first glance) that the correct thing to do is apply aria-required / required properties to the actual `<input />` elements.

http://jqueryvalidation.org/files/demo/radio-checkbox-select-demo.html

Maybe that's outside the scope of this fix though; Adding it to the fieldset would at least assist with screen-reader devices, which was addressed in my original link. :) It must be _somewhere_ on the formfield.

Comment (tractorcow) [inline forms/OptionsetField.php]: Ah, you responded too soon.

Yes, I think you might be right there. :)

**Pull request closed without merging**